### PR TITLE
Http json write series endpoint

### PR DESCRIPTION
**Pull request by corylanou**

### CODE REVIEW

File: handler.go [hunk 1]
```diff
@@ -189,57 +189,23 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, u *User) {
...
-	// Setup HTTP request reader. Wrap in a gzip reader if encoding set in header.
-	reader := r.Body
-	if r.Header.Get("Content-Encoding") == "gzip" {
-		if reader, err = gzip.NewReader(r.Body); err != nil {
-			h.error(w, err.Error(), http.StatusBadRequest)
+	//Read from the request body.
+	dec := json.NewDecoder(r.Body)
+	dec.UseNumber()
+	for {
+		var point Point
```
Comment (pauldix) [inline handler.go]: can we change this to decode an array of points?

File: handler.go [hunk 2]
```diff
@@ -189,57 +189,23 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, u *User) {
 
 // serveWriteSeries receives incoming series data and writes it to the database.
 func (h *Handler) serveWriteSeries(w http.ResponseWriter, r *http.Request, u *User) {
-	// TODO: Authentication.
+	database := r.URL.Query().Get(":db")
+	retentionPolicy := r.URL.Query().Get("retentionPolicy")
```
Comment (pauldix) [inline handler.go]: having the db and the rp in the query string is good for now. But I'm thinking we're going to want a way to write in data via JSON UDP and that would make the formats different between the two.

**Comment by pauldix** [comment]
I had those two line comments, but now that I'm thinking about it, if we want to keep the JSON api for writing points consistent across UDP and HTTP and whatever, we may want to make it a more complex object. Like this:

```go
type Write struct {
  Points []points            `json:"points"`
  Database string          `json:"database"`
  RetentionPolicy string `json:"retentionPolicy"`
}
```

We could even extend that to have shared elements in this top level object. Like this:

```go
type Write struct {
  Points []points             `json:"points"`
  Database string           `json:"database"`
  RetentionPolicy string  `json:"retentionPolicy"`
  Tags map[string]string `json:"tags"`
  Timestamp time.Time  `json:"timestamp"`
}
```

Then if `Timestamp` is supplied, set that on any `Point` that doesn't have it set. And if `Tags` are supplied, have those merge with any `Tags` set on an individual `Point`.

Maybe that last part isn't necessary, but I definitely think the write object would be better.

**Comment by pauldix** [comment]
The HTTP requests would still have to supply username and password in query parameters or basic auth if auth is enabled. The UDP endpoint would bypass authentication because it shouldn't be open to the outside anyway.

**Comment by corylanou** [comment]
@pauldix I can do the later, neither are much of a difference in terms of code, but the later gives more flexibility.  One question on merging tags: if they have the tag defined for both point and top level, should I use what is on the point?

**Comment by pauldix** [comment]
@corylanou: yeah, use the one on the point level. Assume it's more specific so it's an override to the upper level tag.

File: handler.go [hunk 3]
```diff
@@ -189,57 +189,23 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, u *User) {
...
-		h.error(w, err.Error(), http.StatusBadRequest)
-		return
-	}
-
-	// Setup HTTP request reader. Wrap in a gzip reader if encoding set in header.
-	reader := r.Body
-	if r.Header.Get("Content-Encoding") == "gzip" {
-		if reader, err = gzip.NewReader(r.Body); err != nil {
-			h.error(w, err.Error(), http.StatusBadRequest)
+	//Read from the request body.
```
Comment (otoolep) [inline handler.go]: Nit: whitespacing.

File: handler.go (hunk 2 above)
Comment (otoolep) [inline handler.go]: +1 on specifying RP and DB in the JSON body only.

Comment (otoolep) [inline handler.go]: Maybe just the RP, but you get the idea. Keeping the same as much as possible will prevent a lot of confusion.

**Comment by corylanou** [comment]
@pauldix the only final thing we should decide is the route.  Right now it is /db/:db/series but we also pass in database in the json now.  I would propose we just make the route /series to avoid the confusion.

**Comment by pauldix** [comment]
@corylanou: Agree. We should check before the write that the database exists and return a 400 with a good error if it doesn't. I would also suggest maybe just `/write` as the endpoint since the other is `/query`. I think they're a bit more symmetrical, thoughts?

File: handler.go [hunk 4]
```diff
@@ -156,59 +157,71 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, u *User) {
...
 		}
+
+		// TODO corylanou: Check if user can write to specified database
+		//if !user_can_write(br.Database) {
+		//writeError(&Result{Err: fmt.Errorf("%q user is not authorized to write to database %q", u.Name)}, http.StatusUnauthorized)
+		//return
+		//}
+
+		for _, p := range br.Points {
+			if p.Timestamp.IsZero() {
```
Comment (pauldix) [inline handler.go]: what would happen if they actually wanted a zero timestamp?

Comment (corylanou) [inline handler.go]: Good question.  Is that allowed?

Comment (pauldix) [inline handler.go]: Uh, maybe. Let's just punt on it for now.

**Comment by pauldix** [comment]
LGTM, :shipit:

File: handler.go [hunk 5]
```diff
@@ -156,59 +157,71 @@ func (h *Handler) serveQuery(w http.ResponseWriter, r *http.Request, u *User) {
...
-	/* TEMPORARILY REMOVED FOR PROTOBUFS.
-	// Retrieve database from server.
-	db := h.server.Database(r.URL.Query().Get(":db"))
-	if db == nil {
-		h.error(w, ErrDatabaseNotFound.Error(), http.StatusNotFound)
-		return
-	}
+// serveWrite receives incoming series data and writes it to the database.
+func (h *Handler) serveWrite(w http.ResponseWriter, r *http.Request, u *User) {
+	var br batchWrite
```
Comment (otoolep) [inline handler.go]: `br`? Batched request? I was expecting to see it called `bw`. :-)

File: handler_test.go [hunk 6]
```diff
@@ -672,6 +672,72 @@ func TestHandler_AuthenticatedDatabases_UnauthorizedBasicAuth(t *testing.T) {
 	}
 }
```
Comment (otoolep) [inline handler_test.go]: I think we need a test for an empty body.

**Comment by otoolep** [comment]
I know this is merged, but I have a couple of small comments anyway.

File: handler.go (hunk 5 above)
Comment (corylanou) [inline handler.go]: ha!  I'm guessing I changed the type after I had the variable assignment.  I'll make a note to correct this in a future branch.

File: handler_test.go (hunk 6 above)
Comment (corylanou) [inline handler_test.go]: empty body will either blow up the json parser or return an empty struct and validation will fail.  But true, we should have a test for it.  I'll add one.
